src/directed/bfs.rs

```diff
@@ -123,6 +123,126 @@ where
     bfs_core(start, successors, |n| n == start, false)
 }
 
+/// Compute a shortest path using the [breadth-first search
+/// algorithm](https://en.wikipedia.org/wiki/Breadth-first_search) with
+/// [bidirectional search](https://en.wikipedia.org/wiki/Bidirectional_search).
+///
+/// Bidirectional search runs two simultaneous searches: one forward from the start,
+/// and one backward from the end, stopping when the two meet. In many cases this gives
+/// a faster result than searching only in a single direction.
+///
+/// The shortest path starting from `start` up to a node `end` is
+/// computed and returned in a `Some`. If no path can be found, `None`
+/// is returned instead.
+///
+/// - `start` is the starting node.
+/// - `end` is the end node.
+/// - `successors_fn` returns a list of successors for a given node.
+/// - `predecessors_fn` returns a list of predecessors for a given node. For an undirected graph
+///   this will be the same as `successors_fn`, however for a directed graph this will be different.
+///
+/// A node will never be included twice in the path as determined by the `Eq` relationship.
+///
+/// The returned path comprises both the start and end node.
+///
+/// # Example
+///
+/// We will search the shortest path on a chess board to go from (1, 1) to (4, 6) doing only knight
+/// moves.
+///
+/// ```
+/// use pathfinding::prelude::bfs_bidirectional;
+///
+/// static SUCCESSORS: fn(&(i32, i32)) -> Vec<(i32, i32)> = |&(x, y)| vec![
+///     (x+1,y+2), (x+1,y-2), (x-1,y+2), (x-1,y-2),
+///     (x+2,y+1), (x+2,y-1), (x-2,y+1), (x-2,y-1)
+/// ];
+/// let result = bfs_bidirectional(&(1, 1), &(4, 6), SUCCESSORS, SUCCESSORS);
+/// assert_eq!(result.expect("no path found").len(), 5);
+/// ```
+#[allow(clippy::missing_panics_doc)]
+pub fn bfs_bidirectional<N, FNS, FNP, IN>(
```

Comment:
The implementation here is roughly transcribed from pythons networkx https://networkx.org/documentation/stable/_modules/networkx/algorithms/shortest_paths/unweighted.html#bidirectional_shortest_path